tmt/steps/discover/fmf.py

```diff
@@ -403,23 +398,40 @@ def assert_git_url(plan_name: Optional[str] = None) -> None:
 
             # Check what should be extracted from the sources
             if dist_git_extract:
-                if dist_git_extract == '/':
-                    dist_git_extract = sourcedir
+                if len(dist_git_extract) == 1 and dist_git_extract[0] == '/':
+                    # Special case to extract all sources
+                    dist_git_extract = [sourcedir]
                 else:
-                    try:
-                        dist_git_extract = glob.glob(os.path.join(
-                            sourcedir, dist_git_extract.lstrip('/')))[0]
-                    except IndexError:
+                    if any(p == '/' for p in dist_git_extract):
                         raise tmt.utils.DiscoverError(
-                            f"Couldn't glob '{dist_git_extract}' "
-                            f"within extracted sources.")
+                            "Extract pattern '/' cannot be used with other patterns:\n"
+                            f"Requested pattern: {dist_git_extract}")
+                    # Otherwise resolve the glob pattern
+                    # Any string items are glob patterns, otherwise they are dicts with source and destination
+                    glob_patterns = [str(sourcedir.joinpath(p.lstrip('/'))) for p in dist_git_extract if type(p) == str]
+                    # Remove all glob patterns and keep the dict objects
+                    dist_git_extract = [p for p in dist_git_extract if type(p) != str]
+                    for pattern in glob_patterns:
+                        # Add all resolved paths to the paths
+                        try:
+                            dist_git_extract.extend([{
+                                "src": Path(p),
+                                "dest": Path(p).relative_to(sourcedir),
+                            } for p in glob.glob(pattern)])
```

Comment:
This will be a problem: `dist_git_extract` is probably of incompatible type, with `normalize_path_list` it'd be `List[Path]`, therefore adding a dictionary to it would be disallowed. It looks like adding an accumulator variable for these objects would be better, plus also a dataclass to hold these `src`/`dst` pairs, or a simple tuple, `Tuple[Path, Path]` would be also fine. `actual_dist_git_extract: List[Tuple[Path, Path]]` would be a good fit - process `dist_git_extract`, create entries as needed and pushed them into the "normalized" accumulator.

Reply:
Actually, a couple of more suggestions:

* `# Any string items are glob patterns, otherwise they are dicts with source and destination` - this does not seem to be correct, `dist_git_extract` seems to contain strings only, so far, the first dictionary is added to it in the `for pattern in glob_patterns` loop below. It's all strings now - or, better, it should be all `Path` instances after `normalize_path_list()` is applied at the beginning.
* `glob_patterns` is working with strings, but `Path` already offers `glob` support. How about something like this?

```python
for pattern in dist_git_extract:
    for path in sourcedir.glob(str(pattern)):
        actual_dist_git_extract.append((path, path.relative_to(sourcedir)))
```

This should consume patterns (`Path` instances), ask `sourcedir` to list matching paths, and add them all to the new accumulator as `Tuple[Path, Path]`.

Reply:
The key here is that the user can add dict items instead of just strings (see the spec file changes). A tuple would be incompatible with that spec file changes right?

Reply:
I see, then this needs more work, I'm afraid.

* specification now says:

  ```
  dist-git-extract
            directory (glob supported) to copy from extracted sources.
            Set to ``/`` to extract all sources"
   ```
   That does not mention any use of mappings - yes, *schema* has changed, but not the help text or documentation. It's not clear what these `src` and `dst` fields are, whether they are also glob-like or are supposed to be rather precise. Judging by the code, I presume the latter is correct.

* ok, let's say `dist-git-extract` can now accept also a mapping with `src` and `dst` keys - *from fmf*. How would this new type of input look when used on the command line? `--dist-git-extract src:dst`, `--dist-git-extract src=foo,dst=bar`? This needs to be handled too.

Complex input calls for a custom normalization function. The goal is to normalize input to something uniform as soon as possible, when input first enters tmt internals. We must avoid various `isinstance()` or `type()` calls down the stream - if `get('dist-git-extract')` can return a mix of various types, right there and then it must be sorted out into something clearly typed.

Both fmf and CLI must be normalized to the same result, and it looks like a pair of something is the data we're going to work with, hence tuples might be a good choice. I agree, these values are not `Path` instances yet, they are either glob patterns or src/dst pair of paths-to-be. Ok, let's save them as tuples, to preserve the pair, and omit the second member of the pair if there's just single string, to denote the pattern:

```python
def normalize_dist_git_extract(
        key_address: str,
        value: Any,
        logger: tmt.log.Logger) -> List[Tuple[str, Optional[str]]]:
    if value is None:
        return []

    # Only a single string, must be a glob pattern
    if isinstance(value, str):
        return [(value, None)]

    if isinstance(value, (list, tuple)):
        normalized = []

        for item in value:
            if isinstance(item, dict):
                normalized.append((item['src'], item['dst']))

            elif isinstance(item, str):
                # TODO: how to deal with `--dist-git-extract ??`?
                # --dist-git-extract foo:bar => maybe if ':' in item: src, dst = item.split(':'); normalized.append((src, dst) ?

                normalized.append((item, None))

        return normalized

dist_git_extract = normalize_dist_git_extract(self.get('dist-git-extract', []))
```

2. resolve glob patterns, and turn all strings into `Path` instances - at this point, we know there are only paths, patterns would be resolved: walk the normalized input, propagate every complete pair to a new list - please, do not recycle the name of the original list, it would be a different list because at this point patterns and src/dst pairs become `Path` instances -, and resolve glob patterns when only pattern is given, e.g.:

```python
for src, dst in dist_git_extract:
    if dst is None:
        for path in sourcedir.glob(src):
            actual_dist_git_extract.append((path, path.relative_to(sourcedir)))
    else:
        actual_dist_git_extract.append((Path(src), Path(dst))
```

A mix of str and dicts turns into `List[Tuple[str, Optional[str]]]` when we first meet it, then it turns into a `List[Tuple[Path, Path]]` with patterns resolved, to comply with the "`Path` for path-like strings" rule. The transition is explicit, it would handle all kinds of input, and it's clear where a pattern or mapping turns into paths.

Reply:
Thanks for the tips. I've added a quick TODO on the top about the documentation issue. Feel free to update the top comment so that it's more relevant with the current status.

> I see, then this needs more work, I'm afraid.

No worries, it is to be expected with the spec changes

>  It's not clear what these `src` and `dst` fields are, whether they are also glob-like or are supposed to be rather precise. Judging by the code, I presume the latter is correct.

Yes, the latter is correct. I would prefer to allow glob-like path, but how can that be done for `dst`? Is there an equivalent to the regex group matcher that can be used to replace in the glob pattern?

>  How would this new type of input look when used on the command line? `--dist-git-extract src:dst`, `--dist-git-extract src=foo,dst=bar`? This needs to be handled too.

Oh I did not consider the cli. Is there a reference for another dict-like object that is passible in the cli? I think the best is the latter implementation since it would not require any more post-processing? The former would parse it as tuple? The other nuance is how to change the cli to accept list parameters for this? I would have thought there is a mechanism to synchronize it with the provided json-spec.

> Complex input calls for a custom normalization function. The goal is to normalize input to something uniform as soon as possible, when input first enters tmt internals. We must avoid various `isinstance()` or `type()` calls down the stream - if `get('dist-git-extract')` can return a mix of various types, right there and then it must be sorted out into something clearly typed.

Understood. So we need to extract the normalizer to a separate function. Besides the `if string` normalization, is there anything else to be added? The `List[str|dict]` -> `List[dict]` can only happen in this file since it needs the extracted source. 

> Both fmf and CLI must be normalized to the same result, and it looks like a pair of something is the data we're going to work with, hence tuples might be a good choice

You mean the `dict` -> `tuple`/`NamedTuple`? Sure I don't mind that. If the conversion to a NamedTuple can be done automatically on the importer side that would simplify things.

> I agree, these values are not `Path` instances yet, they are either glob patterns or src/dst pair of paths-to-be. Ok, let's save them as tuples, to preserve the pair, and omit the second member of the pair if there's just single string, to denote the pattern:

I think this depends on if we can solve the globing support for `dst`, otherwise the logic of if it's `Path` or a tuple are rather different. I'll have to look at the section again and see if there is any potential confusion. E.g. what if the user only provided `src`, could it be misinterpreted and allowed through despite the ambiguity?

>     2. resolve glob patterns, and turn all strings into `Path` instances - at this point, we know there are only paths, patterns would be resolved: walk the normalized input, propagate every complete pair to a new list - please, do not recycle the name of the original list, it would be a different list because at this point patterns and src/dst pairs become `Path` instances -, and resolve glob patterns when only pattern is given, e.g.:

Good point I'll make a different variable

Reply:
> > It's not clear what these `src` and `dst` fields are, whether they are also glob-like or are supposed to be rather precise. Judging by the code, I presume the latter is correct.
> 
> Yes, the latter is correct. I would prefer to allow glob-like path, but how can that be done for `dst`? Is there an equivalent to the regex group matcher that can be used to replace in the glob pattern?

I suppose the glob pattern is fine, regex pattern would also work, but that can be resolved later.

> 
> > How would this new type of input look when used on the command line? `--dist-git-extract src:dst`, `--dist-git-extract src=foo,dst=bar`? This needs to be handled too.
> 
> Oh I did not consider the cli. Is there a reference for another dict-like object that is passible in the cli? I think the best is the latter implementation since it would not require any more post-processing? The former would parse it as tuple? The other nuance is how to change the cli to accept list parameters for this? I would have thought there is a mechanism to synchronize it with the provided json-spec.

I'm afraid not, there is no out-of-the-box solution (yet). Options given by user would arrive as strings, or rather a list of strings, e.g. `--dist-git-extract foo:bar --dist-git-extract baz` would land in a normalize function as `['foo:bar', 'baz']`. Normalization must take care of all three types of inputs (simple string, dicts from fmf, and possibly src/dst pair in a single string from command line). My example above tries to address all three, although the CLI par form is left commented out as I was not sure how exactly would such an option look like. The important bit is, when normalization is over, we get a uniform list of items, representing the inputs.

> > Complex input calls for a custom normalization function. The goal is to normalize input to something uniform as soon as possible, when input first enters tmt internals. We must avoid various `isinstance()` or `type()` calls down the stream - if `get('dist-git-extract')` can return a mix of various types, right there and then it must be sorted out into something clearly typed.
> 
> Understood. So we need to extract the normalizer to a separate function. Besides the `if string` normalization, is there anything else to be added? The `List[str|dict]` -> `List[dict]` can only happen in this file since it needs the extracted source.

Handling of `--dist-git-extract src:dst` must happen here as well (or whatever the exact form would be, simply the CLI input must be parsed and converted here as well.)


> > Both fmf and CLI must be normalized to the same result, and it looks like a pair of something is the data we're going to work with, hence tuples might be a good choice
> 
> You mean the `dict` -> `tuple`/`NamedTuple`? Sure I don't mind that. If the conversion to a NamedTuple can be done automatically on the importer side that would simplify things.

tmt has no out-of-the-box solution for this conversion, but it would happen in the normalization function, and that's the right place for it.

> > I agree, these values are not `Path` instances yet, they are either glob patterns or src/dst pair of paths-to-be. Ok, let's save them as tuples, to preserve the pair, and omit the second member of the pair if there's just single string, to denote the pattern:
> 
> I think this depends on if we can solve the globing support for `dst`, otherwise the logic of if it's `Path` or a tuple are rather different. I'll have to look at the section again and see if there is any potential confusion. E.g. what if the user only provided `src`, could it be misinterpreted and allowed through despite the ambiguity?

I'd probably go with both or none, if there's just a single key, the normalization function performing the conversion from dicts into something more rigid, tuples or whatever, would raise a `tmt.utils.NormalizationError`, "hey, both keys are needed". But if there's clear semantics and use case for just `src`, it'd be fine too.

> > 2. resolve glob patterns, and turn all strings into `Path` instances - at this point, we know there are only paths, patterns would be resolved: walk the normalized input, propagate every complete pair to a new list - please, do not recycle the name of the original list, it would be a different list because at this point patterns and src/dst pairs become `Path` instances -, and resolve glob patterns when only pattern is given, e.g.:
> > ```
> 
> Good point I'll make a different variable